#!/bin/bash 

#PBS -P jcomvirome
#PBS -N star_array_custom
#PBS -M jmif9945@uni.sydney.edu.au
#PBS -l select=1:ncpus=12:mem=60GB

module load star
module load samtools

# Read in the CSV and extract the specific line for this job
readarray -t myarray < "$mapping_spreadsheet"
export line=${myarray["$PBS_ARRAY_INDEX"]}
export library=$(echo $line | cut -f1 -d,)
export reference=$(echo $line | cut -f2 -d,)
export identity=$(echo $line | cut -f3 -d,)

# Paths
reads_dir="/scratch/$root_project/$project/trimmed_reads/"
reference_dir="/project/$root_project/$project/mapping/reference/"
output_dir="/project/$root_project/$project/mapping/STAR_output/"

# Create output directory if it doesn't exist
mkdir -p "$output_dir"

# Define input files
r1="${reads_dir}${library}_trimmed_R1.fastq.gz"
r2="${reads_dir}${library}_trimmed_R2.fastq.gz"
reference="${reference_dir}${reference}"

# Create a STAR index if not already created (only needs to be done once per reference)
STAR --runThreadN 8 --runMode genomeGenerate --genomeDir "${reference_dir}${library}_index" --genomeFastaFiles $reference

# Run STAR alignment
STAR --runThreadN 8 \
     --genomeDir "${reference_dir}${library}_index" \
     --readFilesIn $r1 $r2 \
     --readFilesCommand zcat \
     --outFileNamePrefix "${output_dir}${library}_" \
     --outSAMtype BAM SortedByCoordinate \
     --outFilterMismatchNoverLmax $identity \
     --outFilterMultimapNmax 10 \
     --outSAMmultNmax 1 \
     --genomeSAindexNbases 5

# Index the BAM file
samtools index "${output_dir}${library}_Aligned.sortedByCoord.out.bam"
samtools view -h "${output_dir}${library}_Aligned.sortedByCoord.out.bam" | grep -v '^@' | cut -f 3 | sort | uniq -c | sort -nr > "${mapping_stats_dir}${library}_num_reads.txt"
samtools depth -a "${output_dir}${library}_Aligned.sortedByCoord.out.bam" | awk '{sum[$1]+=$3; count[$1]++} END {for (ref in sum) print ref, sum[ref]/count[ref]}' > "${mapping_stats_dir}${library}_average_coverage.txt"